<div class="alert alert-block alert-success" style="font-family: Times New Roman">
    <h4><strong>Laboratory Task 3</strong></h4>

<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Perform a forward and backward propagation in python using the inputs from <b>Laboratory Task 2</b>
</p>

```python
x = np.array([1, 0, 1])
y = np.array([1])

# use relu as the activation function.

# learning rate
lr = 0.001
```
</div>

In [1]:
import numpy as np

def relu(z):
    return np.maximum(0, z)

# Inputs
x = np.array([1, 0, 1], dtype=float)   # shape (3,)
y = np.array([1], dtype=float)         # shape (1,)

# Hidden unit weights: 3 inputs -> 2 hidden units
W1 = np.array([
    [0.2, -0.3],
    [0.4,  0.1],
    [-0.5, 0.2]
])

# Output unit weights: 2 hidden units -> 1 output
W2 = np.array([
    [-0.3],
    [-0.2]
])

# Biases/thresholds: theta1, theta2 for hidden; theta3 for output
theta = np.array([-0.4, 0.2, 0.1])

b1 = theta[:2]   # hidden biases
b2 = theta[2]    # output bias

# Forward pass
z1 = x @ W1 + b1
h = relu(z1)

z2 = h @ W2 + b2
y_hat = relu(z2)

# Error
error = (y - y_hat) ** 2
half_error = 0.5 * error

print("z1 =", z1)
print("h =", h)
print("z2 =", z2)
print("y_hat =", y_hat)
print("error =", error)
print("half_error =", half_error)

z1 = [-0.7  0.1]
h = [0.  0.1]
z2 = [0.08]
y_hat = [0.08]
error = [0.8464]
half_error = [0.4232]


In [2]:
import numpy as np

def relu(z):
    return np.maximum(0, z)

def relu_derivative(z):
    return (z > 0).astype(float)

# --- Inputs from Laboratory Task 2 ---
x = np.array([1, 0, 1], dtype=float)   # shape (3,)
y = np.array([1], dtype=float)         # shape (1,)

# Initial hidden weights (3 inputs -> 2 hidden units)
W1 = np.array([
    [0.2, -0.3],
    [0.4,  0.1],
    [-0.5, 0.2]
])  # shape (3, 2)

# Hidden biases
b1 = np.array([-0.4, 0.2])  # shape (2,)

# Initial output weights (2 hidden -> 1 output)
W2 = np.array([
    [-0.3],
    [-0.2]
])  # shape (2, 1)

# Output bias
b2 = np.array([0.1])  # shape (1,)

# Learning rate
lr = 0.001

# ---------- Forward pass ----------
z1 = x @ W1 + b1               # shape (2,)
h = relu(z1)                   # shape (2,)
z2 = h @ W2 + b2               # shape (1,)
y_hat = relu(z2)               # shape (1,)

# Loss (half MSE for a single sample)
loss = 0.5 * np.sum((y - y_hat) ** 2)

# ---------- Backward pass ----------
# Gradient of loss w.r.t. z2
delta2 = (y_hat - y) * relu_derivative(z2)   # shape (1,)

# Gradients for output layer
dW2 = h.reshape(2, 1) @ delta2.reshape(1, 1)  # shape (2, 1)
db2 = delta2                                  # shape (1,)

# Backpropagate to hidden layer
delta1 = (delta2 @ W2.T).flatten() * relu_derivative(z1)  # shape (2,)

# Gradients for hidden layer
dW1 = x.reshape(3, 1) @ delta1.reshape(1, 2)  # shape (3, 2)
db1 = delta1                                  # shape (2,)

# ---------- Update weights ----------
W2_new = W2 - lr * dW2
b2_new = b2 - lr * db2
W1_new = W1 - lr * dW1
b1_new = b1 - lr * db1

# ---------- Output ----------
print("Forward pass:")
print("z1 =", z1)
print("h  =", h)
print("z2 =", z2)
print("y_hat =", y_hat)
print("loss =", loss)
print()
print("Backward pass:")
print("delta2 =", delta2)
print("dW2 =", dW2)
print("db2 =", db2)
print("delta1 =", delta1)
print("dW1 =", dW1)
print("db1 =", db1)
print()
print("Updated weights:")
print("W2_new =\n", W2_new)
print("b2_new =", b2_new)
print("W1_new =\n", W1_new)
print("b1_new =", b1_new)

Forward pass:
z1 = [-0.7  0.1]
h  = [0.  0.1]
z2 = [0.08]
y_hat = [0.08]
loss = 0.4232

Backward pass:
delta2 = [-0.92]
dW2 = [[ 0.   ]
 [-0.092]]
db2 = [-0.92]
delta1 = [0.    0.184]
dW1 = [[0.    0.184]
 [0.    0.   ]
 [0.    0.184]]
db1 = [0.    0.184]

Updated weights:
W2_new =
 [[-0.3     ]
 [-0.199908]]
b2_new = [0.10092]
W1_new =
 [[ 0.2      -0.300184]
 [ 0.4       0.1     ]
 [-0.5       0.199816]]
b1_new = [-0.4       0.199816]


Forward pass:
z1 = [-0.7  0.1]
h  = [0.  0.1]
z2 = [0.08]
y_hat = [0.08]
loss = 0.4232

Backward pass:
delta2 = [-0.92]
dW2 = [[ 0.   ]
 [-0.092]]
db2 = [-0.92]
delta1 = [0.    0.184]
dW1 = [[0.    0.184]
 [0.    0.   ]
 [0.    0.184]]
db1 = [0.    0.184]

Updated weights:
W2_new =
 [[-0.3     ]
 [-0.199908]]
b2_new = [0.10092]
W1_new =
 [[ 0.2      -0.300184]
 [ 0.4       0.1     ]
 [-0.5       0.199816]]
b1_new = [-0.4       0.199816]